enterprise/tools/check_metrics/check_metrics.go

```diff
@@ -0,0 +1,134 @@
+// If the metrics are healthy, this script will return exit code 0 indicating that the rollout should proceed
+// If the metrics are not healthy for the specified amount of time, the script will return an exit code indicating
+// that the release should be rolled back
+//
+// Pre-requisites:
+//   - You must create a tunnel to prometheus and port-forward in order to run this script with
+//     `kubectl --context="" --namespace=<NAMESPACE> port-forward "<POD_NAME>" --address 0.0.0.0 <FROM_PORT>:<TO_PORT>`
+func main() {
+	flag.Parse()
+	err := yaml.PopulateFlagsFromFile(*configPath)
+	if err != nil {
+		log.Fatalf("Could not read metrics config from %s: %s", *configPath, err)
+	}
+	err = log.Configure()
+	if err != nil {
+		log.Fatalf("Could not configure logger: %s", err)
+	}
+	prometheusClient, err := api.NewClient(api.Config{
+		Address: config.PrometheusAddress,
+	})
+	if err != nil {
+		log.Fatalf("Error creating prometheus client: %s\nDid you run the prometheus port-forward command?", err)
+	}
+	promAPI := v1.NewAPI(prometheusClient)
+
+	eg, gctx := errgroup.WithContext(context.Background())
+	for _, metric := range config.PrometheusMetrics {
+		metric := metric
+		eg.Go(func() error {
+			monitoringTimer := time.After(time.Duration(config.MonitoringTimeframeSeconds) * time.Second)
+			monitoringTicker := time.NewTicker(time.Duration(metric.PollingIntervalSeconds) * time.Second)
+			defer monitoringTicker.Stop()
+
+			unhealthyCount := 0
+			for {
+				select {
+				case <-gctx.Done():
+					return nil
+				case <-monitoringTimer:
+					return nil
+				case <-monitoringTicker.C:
+					healthy, err := metricHealthy(promAPI, metric.Name, metric.CanaryQuery, metric.BaselineQuery, metric.HealthThreshold)
+					if err != nil {
+						log.Warningf("Error querying metric %s: %s", metric.Name, err)
+					}
+
+					if healthy {
+						unhealthyCount = 0
+					} else {
+						unhealthyCount++
+					}
+
+					if unhealthyCount >= metric.MaxUnhealthyCount {
+						return errors.New(fmt.Sprintf("%s metrics unhealthy.", metric.Name))
+					}
+				}
+			}
+		})
+	}
+
+	err = eg.Wait()
+	if err != nil {
+		log.Fatalf("Exiting metrics script: %s", err)
+	}
+
+	// If all metrics look healthy for the duration of the monitoring timeframe, continue with rollout
+	os.Exit(metricsHealthyExitCode)
+}
+
+// Returns whether the given metric is healthy compared to the baseline
+// If the metric's success rate is lower than the baseline's by the input threshold, this will return false
+// Otherwise will return true
+func metricHealthy(v1api v1.API, metricName string, query string, baselineQuery string, healthThreshold float64) (bool, error) {
+	ctx := context.Background()
+	log.Debugf("Querying metric %s", metricName)
+
+	result, _, err := v1api.Query(ctx, query, time.Now())
+	if err != nil {
+		return false, err
+	}
+	resultVector := result.(model.Vector)
+	metricValue := model.SampleValue(0)
```

Comment:
sanity check, does the 0'th value correspond to the latest value?

Reply:
The API only returns one value for the current time. If there's data, the vector has size 1. Otherwise it's empty. This line `model.SampleValue(0)` is in case no data is returned, which happens for some of our counter metrics